reconcile/glitchtip/reconciler.py

```diff
@@ -54,6 +62,21 @@ def _reconcile_projects(
                 new_project = Project(name=project.name, platform=project.platform)
             organization_projects.append(new_project)
 
+        for project in [p.desired for p in project_diff.change.values()]:
+            logging.info(
+                ["update_project", organization_slug, project.name, self.client.host]
+            )
+            if not self.dry_run:
+                updated_project = self.client.update_project(
+                    organization_slug=organization_slug,
+                    slug=project.slug,
+                    name=project.name,
+                    platform=project.platform,
+                )
+                organization_projects[
+                    organization_projects.index(project)
+                ] = updated_project
+
         for desired_project in desired_projects:
             current_project = organization_projects[
                 organization_projects.index(desired_project)
```

Comment:
Now that we have `project_diff`, we can directly iterate on current desired pairs:

```python
        for diff_pair in itertools.chain(
            project_diff.change.values(), project_diff.identical.values()
        ):
            current_project = diff_pair.current
            desired_project = diff_pair.desired
```

Reply:
Unfortunately, this won't work. `organization_projects` holds the latest state of all projects, including already applied updates in this reconciliation run. In `project_diff`, we have the differences at the beginning of the `_reconcile_projects` run.

Reply:
In `_reconcile_projects`, changes for `add` and `delete` should not affect this, the only related state change is in `changes` that create new project object.

If we extract the logic of reconcile project teams into a new method:

```python
    def _reconcile_project_teams(self, current_project, desired_project, organization_slug, organization_teams):
        for team in set(current_project.teams).difference(desired_project.teams):
            logging.info(
                [
                    "remove_project_from_team",
                    organization_slug,
                    team.slug,
                    desired_project.slug,
                    self.client.host,
                ]
            )
            del current_project.teams[current_project.teams.index(team)]
            if not self.dry_run:
                self.client.remove_project_from_team(
                    organization_slug=organization_slug,
                    team_slug=team.slug,
                    slug=desired_project.slug,
                )
        for team in set(desired_project.teams).difference(current_project.teams):
            try:
                org_team = organization_teams[organization_teams.index(team)]
            except ValueError:
                logging.error(
                    f"Cannot find team {team.slug} in organization. This shouldn't happen!"
                )
                continue

            logging.info(
                [
                    "add_project_to_team",
                    organization_slug,
                    org_team.slug,
                    current_project.slug,
                    self.client.host,
                ]
            )
            current_project.teams.append(team)
            if not self.dry_run:
                self.client.add_project_to_team(
                    organization_slug=organization_slug,
                    team_slug=org_team.slug,
                    slug=current_project.slug,
                )
```

Then it can be used like:

```python
        for diff_pair in project_diff.change.values():
            project = diff_pair.desired
            logging.info(
                ["update_project", organization_slug, project.name, self.client.host]
            )
            if not self.dry_run:
                updated_project = self.client.update_project(
                    organization_slug=organization_slug,
                    slug=project.slug,
                    name=project.name,
                    platform=project.platform,
                )
                organization_projects[
                    organization_projects.index(project)
                ] = updated_project
                self._reconcile_project_teams(
                    updated_project,
                    project,
                    organization_slug,
                    organization_teams,
                )

        for diff_pair in project_diff.identical.values():
            self._reconcile_project_teams(
                diff_pair.current,
                diff_pair.desired,
                organization_slug,
                organization_teams,
            )
        return organization_projects
```

In this way, `organization_projects` is not used in new method, if we want to keep it for testing purpose, can be simplified by start with an empty list, add projects to it when processing instead of 

```python
            del organization_projects[organization_projects.index(project)]
```

or

```python
                organization_projects[
                    organization_projects.index(project)
                ] = updated_project
```

Reply:
I tried to implement your suggestion, but this would lead to a huge PR. I struggle slightly with the `dry-run` mode and getting the `organization_projects` right. For now, I want to keep it as it is and open a new Jira ticket for the refactoring.

Reply:
The jira ticket: https://issues.redhat.com/browse/APPSRE-8155